### ISSUE TITLE
vim-bundler slows down vim startuptime significantly

### ISSUE THREAD

**Issue by jottr** [tpope/vim-bundler#38]
In larger ruby projects, vim-bundler increases vim startup time to several seconds. 
After bisecting my `.vimrc` I found the vim-bundler plugin to be the culprit. 

`Bundler version 1.6.2`

``` shell
$ rbenv version
2.0.0-p247 (set by RBENV_VERSION environment variable)
```

``` shell
$ vim --version
VIM - Vi IMproved 7.4 (2013 Aug 10, compiled Sep 25 2014 15:27:10)
MacOS X (unix) version
Included patches: 1-430
Compiled by Homebrew
Huge version without GUI.  Features included (+) or not (-):
+acl             +farsi           +mouse_netterm   +syntax
+arabic          +file_in_path    +mouse_sgr       +tag_binary
+autocmd         +find_in_path    -mouse_sysmouse  +tag_old_static
-balloon_eval    +float           +mouse_urxvt     -tag_any_white
-browse          +folding         +mouse_xterm     -tcl
++builtin_terms  -footer          +multi_byte      +terminfo
+byte_offset     +fork()          +multi_lang      +termresponse
+cindent         -gettext         -mzscheme        +textobjects
-clientserver    -hangul_input    +netbeans_intg   +title
+clipboard       +iconv           +path_extra      -toolbar
+cmdline_compl   +insert_expand   +perl            +user_commands
+cmdline_hist    +jumplist        +persistent_undo +vertsplit
+cmdline_info    +keymap          +postscript      +virtualedit
+comments        +langmap         +printer         +visual
+conceal         +libcall         +profile         +visualextra
+cryptv          +linebreak       +python          +viminfo
+cscope          +lispindent      -python3         +vreplace
+cursorbind      +listcmds        +quickfix        +wildignore
+cursorshape     +localmap        +reltime         +wildmenu
+dialog_con      -lua             +rightleft       +windows
+diff            +menu            +ruby            +writebackup
+digraphs        +mksession       +scrollbind      -X11
-dnd             +modify_fname    +signs           -xfontset
-ebcdic          +mouse           +smartindent     -xim
+emacs_tags      -mouseshape      -sniff           -xsmp
+eval            +mouse_dec       +startuptime     -xterm_clipboard
+ex_extra        -mouse_gpm       +statusline      -xterm_save
+extra_search    -mouse_jsbterm   -sun_workshop    -xpm
   system vimrc file: "$VIM/vimrc"
     user vimrc file: "$HOME/.vimrc"
 2nd user vimrc file: "~/.vim/vimrc"
      user exrc file: "$HOME/.exrc"
  fall-back for $VIM: "/usr/local/share/vim"
Compilation: /usr/bin/clang -c -I. -Iproto -DHAVE_CONFIG_H   -F/usr/local/Frameworks -DMACOS_X_UNIX  -Os -w -pipe -march=native -mmacosx-version-min=10.9 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1      
Linking: /usr/bin/clang   -L. -L/Users/jottr/.rbenv/versions/2.0.0-p247/lib  -fstack-protector -L/usr/local/lib -L/usr/local/lib -F/usr/local/Frameworks -Wl,-headerpad_max_install_names -o vim        -lm  -lncurses -liconv -framework Cocoa   -fstack-protector -L/usr/local/lib  -L/System/Library/Perl/5.16/darwin-thread-multi-2level/CORE -lperl -framework Python   -lruby-static -lobjc -L/Users/jottr/.rbenv/versions/2.0.0-p247/lib   

```

**Comment by UmkaDK** [comment]
This one still appears to an issue… Has anyone managed to get to the bottom of this?

**Comment by nikhgupta** [comment]
I am a vim novice, but @tpope do you think this load time difference can be due to `system` calling scripts inside a new shell? The load time on my machine nearly agree with this fact.

What I am trying to say is:

    ± ➲ time ruby -rrbconfig -e "print RbConfig::CONFIG[\"ruby_version\"]"                                                    🕝 [14:17:09] 🔋 [64%]
    2.1.0ruby -rrbconfig -e "print RbConfig::CONFIG[\"ruby_version\"]"  0.06s user 0.03s system 95% cpu 0.097 total

while:

    ± ➲ time zsh -c 'ruby -rrbconfig -e "print RbConfig::CONFIG[\"ruby_version\"]"'                                           🕝 [14:16:49] 🔋 [64%]
    2.1.0zsh -c 'ruby -rrbconfig -e "print RbConfig::CONFIG[\"ruby_version\"]"'  0.38s user 0.17s system 97% cpu 0.569 total

Somehow, the latter coincides with the time diff I am seeing on my startup in Rails projects. (nearly 1.2 seconds). Could this be a factor?

**Comment by nikhgupta** [comment]
I did some tests for this with some settings, and the results can be seen here: https://gist.github.com/nikhgupta/86011b5ab70fb492ec52.

Summarized here are the load times with a minimal vimrc (I did some tests after posting that gist, and hence, some of the following values are not seen in the above gist):

- Without any plugins, without any file: **~60ms**
- Without any plugins, with ruby file: **~195ms**
- With vim-bundler, without any file: **~65ms**
- With vim-bundler, with ruby file: **~470ms**

- Without any plugins, without any file, with 2 `system('echo')` commands in vimrc: **~115ms**
- Without any plugins, with ruby file, with 2 `system('echo')` commands in vimrc: **~255ms**

- Total execution time for 2 commands used by vim-bundler in my shell: **~180ms**
- Without any plugins, without any file, with the 2 vim-bundler `system` calls in vimrc: **~305ms**
- Without any plugins, with ruby file, with the 2 vim-bundler `system` commands in vimrc: **~450ms**

Overall, this all seems consistent to me, and considering that `vim-bundler` uses 2 `system` calls, the above values are as per expected.

@tpope However, I would really really love to know a way to suppress these 2 system calls if possible to lower this down further. One of the commands can be suppressed by setting `$GEM_HOME`, while I am not sure about the other.

**Comment by tpope** [comment]
How long does a single `system('echo')` take? Seems like speeding up your shell might be the low hanging fruit?

**Comment by tpope** [comment]
We already check `$GEM_PATH`; not sure what you hope to accomplish with `$GEM_HOME`.

**Comment by nikhgupta** [comment]
@tpope A single `system('echo')`takes 28ms on my system. And, yeah sorry about that, I meant `$GEM_PATH` itself.

Also, by checking the time for these commands, what I wanted to convey was that `vim-bundler` is not slow, but instead command invocations via `system` are inherently slow for users who are having such problems.

**Comment by tpope** [comment]
Yeah no doubt that even in the best case `system()` is going to be slow. I was guessing yours was *very* slow but mine is 20ms so maybe not.

See #34 for more info about that `abi_version` assignment, and about the challenges of doing it without a `system()`. If you can figure out an improvement there, great.

Looking at the code also suggests that maybe `abi_version` isn't even needed in the common case? If true, it could probably be made lazy.

**Comment by Confusion** [comment]
You will also run into vim starting noticeably slowly when using vim-bundler in combination with jruby. 

The abi_version system call mentioned in [an earlier comment](https://github.com/tpope/vim-bundler/issues/38#issuecomment-60169415) will each take ~2 seconds, due to JVM startup time. Unfortunately I don't see a reliable alternative either. 

What helps a bit for the 'abi_version' call, is adding the `--disable-gems` switch to the ruby call. This also speeds up the command for regular MRI ruby (from ~50 ms to ~30 ms on my machine). Jruby users could further ensure JRUBY_OPTS="--dev" is set. Those two combined reduce the time to ~1 second.

In my setup the abi_version call is actually performed twice during startup and I'm not yet clear on why that is.

**Comment by tpope** [comment]
With `RUBY VERSION` in the lock file now, maybe we could lean on that? What does it look like for JRuby?

**Comment by Confusion** [comment]
For instance like `ruby 2.3.3p0 (jruby 9.1.17.0)`; indicating compatibility with MRI 2.3.3. So no abi version unfortunately.

**Comment by tpope** [comment]
Are there any exceptions to the rule that Ruby version 2.x.y has ABI version 2.x.0? It *seems* like it would be simple to map between the two. Maybe limit ourselves to 2.4 and earlier so we don't get screwed by future releases.

I guess one additional concern is that `RUBY VERSION` might not reflect the actually running version, seeing as how it's a basically a comment.

**Comment by rainerborene** [comment]
Maybe we can do this Ruby detection in background using `jobstart`?

**Comment by tpope** [comment]
We need the resulting path synchronously for most uses, so the most that could do would be to allow it to happen in parallel with the rest of the detection. And it doesn't sound like that would help much.

**Comment by tpope** [comment]
It feels worth calling out that 04d2980e04c745744f18573213631ab897d6aff5 combined the ABI version query and Gem path query into a single call. Multiple comments of mine above suggest possible ways to avoid the ABI version query, but they won't help with the Gem path query, so they're effectively useless now as far as I'm concerned.
